data_model/src/asset.rs

```diff
@@ -384,6 +385,27 @@ pub struct DefinitionId {
     pub domain_id: <Domain as Identifiable>::Id,
 }
 
+/// Asset Definition Identification is represented by `name#domain_name` string.
+impl FromStr for DefinitionId {
+    type Err = ParseError;
+
+    fn from_str(string: &str) -> Result<Self, Self::Err> {
+        let mut split = string.split('#');
+        match (split.next(), split.next(), split.next()) {
+            (Some(""), _, _) => Err(ParseError {
+                reason: "Asset Definition ID cannot be empty",
+            }),
+            (Some(name), Some(domain_id), None) if !domain_id.is_empty() => Ok(Self {
+                name: name.parse()?,
+                domain_id: domain_id.parse()?,
+            }),
+            _ => Err(ParseError {
+                reason: "Asset Definition ID should have format `asset#domain`",
```

Comment:
Ideally we want to retain the candidate `String` as well. Haskell has a lot of resources on the benefits of functional error handling (where you don't just throw away the partial candidate). 

Ideally this should allow people to fix their `String` representations.